retrieval/discovery/azure.go

```diff
@@ -0,0 +1,214 @@
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ad.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// azureClient represents multiple Azure Resource Manager providers.
+type azureClient struct {
+	nic network.InterfacesClient
+	vm  compute.VirtualMachinesClient
+}
+
+// createAzureClient is a helper function for creating an Azure compute client to ARM.
+func createAzureClient(cfg config.AzureSDConfig) (azureClient, error) {
+	var c azureClient
+	oauthConfig, err := azure.PublicCloud.OAuthConfigForTenant(cfg.TenantID)
+	if err != nil {
+		return azureClient{}, err
+	}
+	spt, err := azure.NewServicePrincipalToken(*oauthConfig, cfg.ClientID, cfg.ClientSecret, azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return azureClient{}, err
+	}
+
+	c.vm = compute.NewVirtualMachinesClient(cfg.SubscriptionID)
+	c.vm.Authorizer = spt
+
+	c.nic = network.NewInterfacesClient(cfg.SubscriptionID)
+	c.nic.Authorizer = spt
+
+	return c, nil
+}
+
+func (ad *AzureDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{}
+	client, err := createAzureClient(*ad.cfg)
+	if err != nil {
+		return tg, fmt.Errorf("could not create Azure client: %s", err)
+	}
+
+	var machines []compute.VirtualMachine
+	result, err := client.vm.ListAll()
+	if err != nil {
+		return tg, fmt.Errorf("could not list virtual machines: %s", err)
+	}
+	machines = append(machines, *result.Value...)
+
+	// If we still have results, keep going until we have no more.
+	for result.NextLink != nil {
+		result, err = client.vm.ListAllNextResults(result)
+		if err != nil {
+			return tg, fmt.Errorf("could not list virtual machines: %s", err)
+		}
+		machines = append(machines, *result.Value...)
+	}
+	log.Debugf("Found %d virtual machines during Azure discovery.", len(machines))
+
+	// We have the slice of machines. Now turn them into targets.
+	// Doing them in go routines because the network interface calls are slow.
+	type target struct {
+		labelSet model.LabelSet
+		err      error
+	}
+
+	ch := make(chan target, len(machines))
+	for i, vm := range machines {
+		go func(i int, vm compute.VirtualMachine) {
+			resourceGroup := strings.ToLower(strings.Split(*vm.ID, "/")[4])
+			labels := model.LabelSet{
+				azureLabelMachineID:            model.LabelValue(*vm.ID),
+				azureLabelMachineName:          model.LabelValue(*vm.Name),
+				azureLabelMachineLocation:      model.LabelValue(*vm.Location),
+				azureLabelMachineResourceGroup: model.LabelValue(resourceGroup),
+			}
+
+			if vm.Tags != nil {
+				for k, v := range *vm.Tags {
+					name := strutil.SanitizeLabelName(k)
+					labels[azureLabelMachineTag+model.LabelName(name)] = model.LabelValue(*v)
+				}
+			}
+
+			// Get the IP address information via seperate call to the network provider.
+			for _, nic := range *vm.Properties.NetworkProfile.NetworkInterfaces {
```

Comment:
Yes, basically the comment in 185 still holds.

What I mean here is:
* in the loop in L204 we expect exactly one message per machine (be that an actual target, an error, or no-op as the comment in L185 suggests)
* In  L149 we start one goroutine for each machine. This goroutine is responsible for sending exactly one target on `ch`, because we read exactly one per machine.
* In L167 though, we start a loop in each of these goroutines. We loop over all network interfaces for this machine. 
 * In L174 there may be an error, we send an error message on `ch` and abort. That's cool.
 * In L199 we had success, we have populated the labelset with the network address and send a target on `ch`. Now we continue with the next interface. This is where I'm confused. 
   * a) in the next iteration we send a new error or target on the channel. Now we have two messages send to `ch` for one machine. But the reading loop expects exactly one per target.
    * b) in the next iteration we write into the same labelset. A labelset is only a reference though, so we modify the contents of a previous message.

My question is what we want as the appropriate behavior.

* Do we just want the first network interface that succeeds? Then we can just but a `break` at the end of the loop.
* Do we want all interfaces to be available somehow? Then we could take the first one for the `__address__` label and `azureLabelMachinePrivateIP` from the first one and add subsequent ones as numbered meta labels. We do something similar for "additional endpoints" in [serversets](https://github.com/prometheus/prometheus/blob/master/retrieval/discovery/serverset.go#L151-L158).

Do my concerns make sense? Am I missing something?